templates/nodeovc/nodeovc.py

```diff
@@ -0,0 +1,248 @@
+        # check if port 22 is already created or requested
+        ssh_present = ('22' in existent_ports) or ('22' in requested_ports)
+
+        # if port 22 is not created, add to requested
+        if not ssh_present:
+            ssh_port = {'source':'22', 'destination': '22'}
+            port_forwards.append(ssh_port)
+
+        # delete ports that are not requested
+        for port in existent_ports:
+            if port != '22' or port not in requested_ports:
+                machine.portforward_delete(port)
+
+        # add requested ports
+        for port in port_forwards:
+            # check if ports do not exist yet
+            if port['destination'] not in existent_ports:
+                # create portforward
+                machine.portforward_create(
+                    publicport=port['destination'], 
+                    localport=port['source'], 
+                    protocol='tcp',
+                    )
+
+    def start(self):
+        """ Start the VM """
+
+        machine = self.machine
+        if machine:
+            machine.start()
+        else:
+            self.err_log_machine_not_found()
+
+    def stop(self):
+        """ Stop the VM """
+
+        machine = self.machine
+        if machine:
+            machine.stop()
+        else:
+            self.err_log_machine_not_found()
+
+    def restart(self):
+        """ Restart the VM """
+
+        machine = self.machine
+        if machine:
+            machine.restart()
+        else:
+            self.err_log_machine_not_found()
+
+    def pause(self):
+        """ Pause the VM """
+
+        machine = self.machine
+        if machine:
+            machine.pause()
+        else:
+            self.err_log_machine_not_found()
+
+    def resume(self):
+        """ Resume the VM """
+
+        machine = self.machine
+        if machine:
+            machine.resume()
+        else:
+            self.err_log_machine_not_found()
+
+    def reset(self):
+        """ Reset the VM """
+
+        machine = self.machine
+        if machine:
+            machine.reset()    
+        else:
+            self.err_log_machine_not_found()      
+
+    def snapshot(self):
+        """
+        Action that creates a snapshot of the machine
+        """
+        machine = self.machine
+        if machine:
+            machine.snapshot_create()
+        else:
+            self.err_log_machine_not_found()
+
+    def snapshot_rollback(self, snapshot_epoch=None):
+        """
+        Action that rolls back the machine to a snapshot
+        """    
+        machine = self.machine
+        if machine and snapshot_epoch:       
+            machine.snapshot_rollback(snapshot_epoch)
+            machine.start()
+        else:
+            self.err_log_machine_not_found()
+
+    def snapshot_delete(self,  snapshot_epoch=None):
+        """
+        Action that deletes a snapshot of the machine
+        """
+        machine = self.machine
+        if machine and snapshot_epoch:       
```

Comment:
Here you can log that a machine doesn't exsit while it does because of `if machine and snapshot_epoch`, this condition will also fail is snapshot_epoch is None.
I don't think you should even have `snapshot_epoch=None` this argument is required, not optional.
Also do check for machine existing and is snapshot_epoch is something valid and raise or log proper error accordingly